azkaban-common/src/main/java/azkaban/user/UserUtils.java

```diff
@@ -100,38 +101,39 @@ static void setupWatch(final Map<String, ParseConfigFile> configFileMap) throws
       log.warn("Watchservice was not setup for any config file(s).");
       try {
         watchService.close();
-      } catch (IOException e) {
+      } catch (final IOException e) {
         log.warn("IOException while closing watchService. " + e.getMessage());
       }
       return;
     }
 
-    Runnable runnable = () -> {
+    final Runnable runnable = () -> {
       // Watchservice is established, now listen for the events till eternity!
-      for (;; ) {
-        WatchKey watchKey;
+      for (; ; ) {
+        final WatchKey watchKey;
         try {
           watchKey = watchService.take();
-        } catch (InterruptedException ie) {
+          // Wait for a second to ensure there is only one event for a modification.
+          // WatchService otherwise creates two events 1 for content and 1 for modification time.
+          Thread.sleep(1000L);
```

Comment:
Really? Does it matter how "often" `watchService.take()` is called, or are all events returned by `watchKey.pollEvents()` any way?

Reply:
All events in the queue are returned by take(). Please take a look at the latest patch. Hopefully it clarifies.
https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-the-same-event

Reply:
Ok. Sorry I didn't expect the WatchService  API to work like that. As far as I understood, this kind of scenario is still possible:

> user: 0.0 s: start writing, file is empty
watcher thread: 0.01 s: `watchService.take()` returns, go to `Thread.sleep`
user: 0.1 s: write rest of the stuff to file & close it
watcher thread: 1.01 s: `Thread.sleep` has ended, `watchKey.pollEvents()` is called
user: 1.02s: start writing again, file is empty again
watcher thread: 1.03 s: **try to read the file -> fail because it's empty**, as long as exception is caught, **loop starts from beginning** -> go to `watchService.take()` again
user: 1.1 s: write rest of the stuff to file & close it
watcher thread: 1.11 s: `watchService.take()` returns, go to `Thread.sleep`
watcher thread: 2.11 s: Thread.sleep has ended, `watchKey.pollEvents()` is called
watcher thread: 2.13 s: **reading file succeeds this time**

I guess it's still good to try to avoid the failure when possible, but this kind of corner cases still exist. Should be fine I guess, but makes me wonder if it would be possible to lock the file to make it 100% robust :)

Reply:
@juhoautio thanks for reviewing the code. As you mentioned, incase the sleep is not enough, the exception would occur and eventually another event will catch the write, thus guaranteeing that reload occurs. Did you get chance to look at the latest patch? I took some cues from your work.

Reply:
Regarding locking, we don't do it now. Maybe we can create an issue to track it separately?